lib/mllparty/connection_hub/client_monitor.ex

```diff
@@ -0,0 +1,160 @@
+    {:noreply, new_state}
+  end
+
+  def handle_connection_up(endpoint, state) do
+    # Get current timestamp
+    now = DateTime.utc_now()
+    # Connection is up, check if it was down previously
+    if state.down_at do
+      seconds = DateTime.diff(now, state.down_at)
+
+      message =
+        "#{endpoint} connection is up. Last down duration: #{TimeUtils.format_seconds(seconds)}."
+
+      Logger.info(message)
+      # If a downtime notification was sent, follow up that it's connected again
+      if state.down_notification_sent do
+        SlackNotification.send_notification(message)
+      end
+    end
+
+    # Reset down_at, down_notification_sent, reconnect_attempted
+    %{down_at: nil, down_notification_sent: false, reconnect_attempted: false}
+  end
+
+  def handle_connection_down(endpoint, connected, pending_reconnect, state) do
+    # Get current timestamp
+    now = DateTime.utc_now()
+    # Connection is down
+    new_state =
+      if state.down_at == nil do
+        message =
+          case connected do
+            false ->
+              "#{endpoint} connection is down"
+
+            {:unavailable, reason} ->
+              "#{endpoint} connection is down (mllp client connection status unavailable, reason: #{reason})"
+          end
+
+        Logger.info(message)
+        %{down_at: now, down_notification_sent: false}
+      else
+        # The presence of down_at indicates the connection has been down for a while
+        # Calculate the downtime duration
+        duration = DateTime.diff(now, state.down_at)
+
+        # If duration has been 60 seconds or more, send a slack notification (unless it was already sent)
+        sent =
+          if duration >= 60 and not state.down_notification_sent do
+            message =
+              case connected do
+                false ->
+                  "#{endpoint} connection has been down for #{duration} seconds"
+
+                {:unavailable, reason} ->
+                  "#{endpoint} connection has been down for #{duration} seconds (mllp client connection status unavailable, reason: #{reason})"
+              end
+
+            Logger.warn(message)
+            SlackNotification.send_notification(message)
+            true
+          else
+            state.down_notification_sent
+          end
+
+        %{down_at: state.down_at, down_notification_sent: sent}
+      end
+
+    # Try to reconnect if the client isn't pending reconnect and we haven't already attempted to reconnect
+    reconnect_attempted =
+      if pending_reconnect == false && state.reconnect_attempted == false do
+        client_pid = ClientWrapper.child_client_pid(state.client_wrapper_pid)
+
+        try do
+          Logger.info("Attempting reconnect for #{endpoint}")
+
+          # This method instructs the client to reconnect using it's configured backoff or auto reconnect interval
+          MLLP.Client.reconnect(client_pid)
+        catch
+          :exit, {reason, _} ->
```

Comment:
Are these catches still necessary? Or were they an artifact of some of those initial spikes?

Reply:
I believe they are necessary to avoid the monitor crashing while waiting for the MLLP.Client.reconnect to respond